Speed up closing all FDs

DaniPopes opened the pull request:
Use `libc::close_from` where available (Linux/Android, FreeBSD) to speed up closing all open FDs before execve in the child process.

Fixes #19.

DaniPopes committed:
Speed up closing all FDs

Use `libc::close_from` where available (Linux/Android, FreeBSD)
to speed up closing all open FDs before execve in the child process.

Fixes #19.

DaniPopes commented on src/lib.rs:
```diff
@@ -719,17 +717,35 @@ fn make_controlling_tty(ptm: &Master) -> Result<()> {
+    // Close `0..n` normally.
+    close_descriptors_iter(num_known_fds, except)?;
+    // Close `n..` with one syscall.
+    let res = unsafe { nix::libc::close_range(num_known_fds as std::ffi::c_uint, !0, 0) };
```
On Linux this is only available since glibc 2.34 so it's not in `nix`: https://github.com/nix-rust/nix/issues/2538
Do you think this is a problem for including Linux/Android in the cfg?

zhiburt replied:
> On Linux this is only available since glibc 2.34

In my opinion it is a problem.
Because it's also available only after linux kernel 5.0+ 

And in rust we can't yet have conditional compilation for such things :(.
Theoretically we could have something in `build.rs`.
But I am not sure it's necessary yet.

So I would suggest to add a feature flag for this.
Since it will compile not on all platforms.

I don't know what the name of it have to be though.

> Do you think this is a problem for including Linux/Android in the cfg?

Apparently there's no such function on android yet.

https://docs.rs/libc/latest/aarch64-linux-android/libc/index.html?search=close_range

zhiburt reviewed (commented):
Hey @DaniPopes 

Forgive me....
Just got busy sometimes......

This looks good to me.

The only note:
We could optimize `close_all_descriptors`.

In your implementation we `close_range` from the last descriptor we given.
But we could close all ranges in between the given except list (need to sort list beforehand).

```
list: [0, 1, 2, 10, 30]

close [0, 1] -> close []
close [1, 2] -> close []
close [2, 10] -> close [3, 4, 5, 6, 7, 8, 9]
close [10, 30] -> close [11, 12, 13, ..., 29]
close [31, SysconfVar::OPEN_MAX] -> close [32, 33, ....]
```

It's not a big deal,
But potentially brings a bit improvement.

Let me know what you think about my comments.

In anycase good job.
I'll merge it in case if you won't answer.
And will try to release tomorrow.

Once again sorry for the delay

DaniPopes commented:
No worries at all! I've added a "close-range" feature, removed android from the supported list, and updated the implementation to iterate over ranges.

zhiburt commented on Cargo.toml:
```diff
@@ -14,3 +14,8 @@ categories = ["development-tools", "command-line-interface"]
+[features]
+# Use `close_range` on supported platforms to improve performance when closing many descriptors.
+# This is only available since Linux 5.9 and glibc 2.34, and is always enabled on FreeBSD.
+close-range = []
```
I like the name of it :)

zhiburt replied:
Just a historic note:

> always enabled on FreeBSD.

Seems like only after FreeBSD 8.0 - which is old enough - so good enough to keep it this way

zhiburt reviewed (approved):
Looks good to me.

Thanks a lot for your work.
I'll try to make a release later this day. (need to check whether I got some todos here)
